collector/indices_mappings.go

```diff
@@ -0,0 +1,182 @@
+package collector
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	defaultIndicesMappingsLabels = []string{"index"}
+)
+
+type indicesMappingsMetric struct {
+	Type  prometheus.ValueType
+	Desc  *prometheus.Desc
+	Value func(indexMapping IndexMapping) float64
+}
+
+// IndicesMappings information struct
+type IndicesMappings struct {
+	logger log.Logger
+	client *http.Client
+	url    *url.URL
+
+	up                              prometheus.Gauge
+	totalScrapes, jsonParseFailures prometheus.Counter
+
+	metrics []*indicesMappingsMetric
+}
+
+// NewIndicesMappings defines Indices Mappings Prometheus metrics
+func NewIndicesMappings(logger log.Logger, client *http.Client, url *url.URL) *IndicesMappings {
+	subsystem := "indices_mappings_stats"
+
+	return &IndicesMappings{
+		logger: logger,
+		client: client,
+		url:    url,
+
+		up: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "up"),
+			Help: "Was the last scrape of the ElasticSearch Indices Mappings endpoint successful.",
+		}),
+		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "total_scrapes"),
+			Help: "Current total ElasticSearch Indices Mappings scrapes.",
+		}),
+		jsonParseFailures: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, subsystem, "json_parse_failures"),
+			Help: "Number of errors while parsing JSON.",
+		}),
+		metrics: []*indicesMappingsMetric{
+			{
+				Type: prometheus.GaugeValue,
+				Desc: prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, subsystem, "number_of_fields"),
+					"Current number fields within cluster.",
+					defaultIndicesMappingsLabels, nil,
+				),
+				Value: func(indexMapping IndexMapping) float64 {
+					return countFieldsRecursive(indexMapping.Mappings.Properties, 0)
+				},
+			},
+		},
+	}
+}
+
+func countFieldsRecursive(properties Properties, fieldCounter float64) float64 {
```

Comment:
This probably makes sense as a pointer to reduce the number of memory allocations for each recursion depth. That would also mean that there is no return necessary.